// firebaseui-angular/index.js
import { Component, EventEmitter, Injectable, NgModule, Output } from '@angular/core';
import { CommonModule } from '@angular/common';
import { auth } from 'firebaseui';
import * as firebaseui from 'firebaseui';
import { AngularFireAuth } from 'angularfire2/auth';
import { auth as auth$1 } from 'firebase/app';
import * as firebase from 'firebase/app';
import 'rxjs/Rx';

var FirebaseUIService = (function () {
    /**
     * @param {?} angularFireAuth
     */
    function FirebaseUIService(angularFireAuth) {
        // store the firebaseui instance on the window object to prevent double initialization
        if (!window.firebaseUiInstance) {
            window.firebaseUiInstance = new auth.AuthUI(angularFireAuth.auth);
        }
        this.firebaseUiInstance = window.firebaseUiInstance;
    }
    return FirebaseUIService;
}());
FirebaseUIService.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
FirebaseUIService.ctorParameters = function () { return [
    { type: AngularFireAuth, },
]; };

var FirebaseUIAuthConfig = (function () {
    function FirebaseUIAuthConfig() {
    }
    return FirebaseUIAuthConfig;
}());
var FirebaseUISignInSuccess = (function () {
    function FirebaseUISignInSuccess() {
    }
    return FirebaseUISignInSuccess;
}());
var CredentialHelper = {};
CredentialHelper.AccountChooser = 0;
CredentialHelper.None = 1;
CredentialHelper[CredentialHelper.AccountChooser] = "AccountChooser";
CredentialHelper[CredentialHelper.None] = "None";
var AuthProvider = {};
AuthProvider.Google = 0;
AuthProvider.Facebook = 1;
AuthProvider.Twitter = 2;
AuthProvider.Github = 3;
AuthProvider.Password = 4;
AuthProvider.Phone = 5;
AuthProvider[AuthProvider.Google] = "Google";
AuthProvider[AuthProvider.Facebook] = "Facebook";
AuthProvider[AuthProvider.Twitter] = "Twitter";
AuthProvider[AuthProvider.Github] = "Github";
AuthProvider[AuthProvider.Password] = "Password";
AuthProvider[AuthProvider.Phone] = "Phone";
var AuthProviders = {};
AuthProviders.Google = 0;
AuthProviders.Facebook = 1;
AuthProviders.Twitter = 2;
AuthProviders.Github = 3;
AuthProviders.Password = 4;
AuthProviders.Phone = 5;
AuthProviders[AuthProviders.Google] = "Google";
AuthProviders[AuthProviders.Facebook] = "Facebook";
AuthProviders[AuthProviders.Twitter] = "Twitter";
AuthProviders[AuthProviders.Github] = "Github";
AuthProviders[AuthProviders.Password] = "Password";
AuthProviders[AuthProviders.Phone] = "Phone";
var AuthMethods = {};
AuthMethods.Popup = 0;
AuthMethods.Redirect = 1;
AuthMethods[AuthMethods.Popup] = "Popup";
AuthMethods[AuthMethods.Redirect] = "Redirect";

var FirebaseUIComponent = (function () {
    /**
     * @param {?} angularFireAuth
     * @param {?} firebaseUiConfig
     * @param {?} firebaseUIService
     */
    function FirebaseUIComponent(angularFireAuth, firebaseUiConfig, firebaseUIService) {
        this.angularFireAuth = angularFireAuth;
        this.firebaseUiConfig = firebaseUiConfig;
        this.firebaseUIService = firebaseUIService;
        this.signInSuccessCallback = new EventEmitter();
    }
    /**
     * @param {?} provider
     * @return {?}
     */
    FirebaseUIComponent.getAuthProvider = function (provider) {
        switch (provider) {
            case AuthProvider.Google:
                return auth$1.GoogleAuthProvider.PROVIDER_ID;
            case AuthProvider.Facebook:
                return auth$1.FacebookAuthProvider.PROVIDER_ID;
            case AuthProvider.Twitter:
                return auth$1.TwitterAuthProvider.PROVIDER_ID;
            case AuthProvider.Github:
                return auth$1.GithubAuthProvider.PROVIDER_ID;
            case AuthProvider.Password:
                return auth$1.EmailAuthProvider.PROVIDER_ID;
            case AuthProvider.Phone:
                return auth$1.PhoneAuthProvider.PROVIDER_ID;
        }
    };
    /**
     * @return {?}
     */
    FirebaseUIComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.angularFireAuth.authState.subscribe(function (value) {
            if (!value) {
                if (_this.firebaseUiConfig.providers.length !== 0) {
                    _this.firebaseUIPopup();
                }
                else {
                    throw new Error('There must be at least one AuthProvider.');
                }
            }
        });
    };
    /**
     * @return {?}
     */
    FirebaseUIComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    /**
     * @param {?} authConfig
     * @return {?}
     */
    FirebaseUIComponent.prototype.getUIAuthConfig = function (authConfig) {
        var _this = this;
        var /** @type {?} */ authProviders = [];
        for (var _i = 0, _a = authConfig.providers; _i < _a.length; _i++) {
            var provider = _a[_i];
            if (!!((provider)).customConfig) {
                provider = ((provider));
                var /** @type {?} */ providerWithConfig = provider.customConfig;
                providerWithConfig['provider'] = FirebaseUIComponent.getAuthProvider(provider.provider);
                authProviders.push(providerWithConfig);
            }
            else {
                authProviders.push(FirebaseUIComponent.getAuthProvider(/** @type {?} */ (provider)));
            }
        }
        var /** @type {?} */ tosURL = authConfig.tos ? authConfig.tos : '';
        var /** @type {?} */ authMethod = 'popup';
        switch (authConfig.method) {
            case null:
            case AuthMethods.Popup:
                break;
            case AuthMethods.Redirect:
                authMethod = 'redirect';
                break;
            default:
                throw new Error("Unknown auth method. Valid: [AuthMethods.Popup, AuthMethods.Redirect]");
        }
        var /** @type {?} */ credentialHelper;
        switch (authConfig.credentialHelper) {
            case CredentialHelper.None:
                credentialHelper = auth.CredentialHelper.NONE;
                break;
            case CredentialHelper.AccountChooser:
            default:
                credentialHelper = auth.CredentialHelper.ACCOUNT_CHOOSER_COM;
                break;
        }
        var /** @type {?} */ nativeConfiguration = {
            callbacks: {
                signInSuccess: function (currentUser, credential, redirectUrl) {
                    _this.signInSuccessCallback.emit({
                        currentUser: currentUser,
                        credential: credential,
                        redirectUrl: redirectUrl
                    });
                    return !!authConfig.signInSuccessUrl;
                }
            },
            signInFlow: authMethod,
            signInOptions: authProviders,
            tosUrl: tosURL,
            credentialHelper: credentialHelper
        };
        if (!!authConfig.signInSuccessUrl) {
            nativeConfiguration.signInSuccessUrl = authConfig.signInSuccessUrl;
        }
        return nativeConfiguration;
    };
    /**
     * @return {?}
     */
    FirebaseUIComponent.prototype.firebaseUIPopup = function () {
        var /** @type {?} */ firebaseUiInstance = this.firebaseUIService.firebaseUiInstance;
        firebaseUiInstance.start('#firebaseui-auth-container', this.getUIAuthConfig(this.firebaseUiConfig));
    };
    return FirebaseUIComponent;
}());
FirebaseUIComponent.decorators = [
    { type: Component, args: [{
                selector: 'firebase-ui',
                template: "\n        <div id=\"firebaseui-auth-container\"></div>"
            },] },
];
/**
 * @nocollapse
 */
FirebaseUIComponent.ctorParameters = function () { return [
    { type: AngularFireAuth, },
    { type: FirebaseUIAuthConfig, },
    { type: FirebaseUIService, },
]; };
FirebaseUIComponent.propDecorators = {
    'signInSuccessCallback': [{ type: Output, args: ['signInSuccess',] },],
};

var FirebaseUIModule = (function () {
    function FirebaseUIModule() {
    }
    /**
     * @param {?} firebaseUiAuthConfig
     * @return {?}
     */
    FirebaseUIModule.forRoot = function (firebaseUiAuthConfig) {
        return {
            ngModule: FirebaseUIModule,
            providers: [
                { provide: FirebaseUIAuthConfig, useValue: firebaseUiAuthConfig }
            ]
        };
    };
    return FirebaseUIModule;
}());
FirebaseUIModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    FirebaseUIComponent,
                ],
                exports: [
                    FirebaseUIComponent,
                ],
                providers: [
                    FirebaseUIService
                ]
            },] },
];
/**
 * @nocollapse
 */
FirebaseUIModule.ctorParameters = function () { return []; };

export { FirebaseUIModule, FirebaseUIComponent, FirebaseUIService, FirebaseUIAuthConfig, FirebaseUISignInSuccess, CredentialHelper, AuthProvider, AuthProviders, AuthMethods };
